validator/c/src/lib.rs

```diff
@@ -0,0 +1,609 @@
+// Functions dereferencing raw pointers are kind of par for the course in a C
+// interface.
+#![allow(clippy::not_unsafe_ptr_arg_deref)]
+
+use std::cell::RefCell;
+
+thread_local! {
+    /// Most recent error message, stored in thread-local storage for
+    /// thread-safety.
+    pub static LAST_ERROR: RefCell<std::ffi::CString> = RefCell::new(std::ffi::CString::default());
+}
+
+/// Pushes an error message.
+fn set_last_error<S: AsRef<str>>(s: S) {
+    LAST_ERROR.with(|f| {
+        *f.borrow_mut() = std::ffi::CString::new(s.as_ref()).unwrap_or_default();
+    });
+}
+
+/// Returns the most recent error message. Note that the returned pointer is
+/// only valid until the next call that the current thread makes to this
+/// library.
+#[no_mangle]
+pub extern "C" fn substrait_validator_get_last_error() -> *const libc::c_char {
+    LAST_ERROR.with(|f| {
+        let reference = f.borrow();
+        reference.as_bytes_with_nul().as_ptr() as *const libc::c_char
+    })
+}
+
+/// Parser/validator configuration handle.
+pub struct ConfigHandle {
+    pub config: substrait_validator_core::Config,
+}
+
+/// Creates a parser/validator configuration structure.
+#[no_mangle]
+pub extern "C" fn substrait_validator_config_new() -> *mut ConfigHandle {
+    // Create a box to store the return value handle on the stack.
+    let handle = Box::new(ConfigHandle {
+        config: substrait_validator_core::Config::new(),
+    });
+
+    // Convert the box to its raw pointer and relinquish ownership.
+    Box::into_raw(handle)
+}
+
+/// Instructs the validator to ignore protobuf fields that it doesn't know
+/// about yet (i.e., that have been added to the Substrait protobuf
+/// descriptions, but haven't yet been implemented in the validator) if the
+/// fields are set to their default value. If this option isn't set, or if an
+/// unknown field is not set to its default value, a warning is emitted.
+///
+/// Returns whether the function was successful. If false is returned, retrieve
+/// the error message with substrait_validator_get_last_error().
+#[no_mangle]
+pub extern "C" fn substrait_validator_config_ignore_unknown_fields_set_to_default(
+    config: *mut ConfigHandle,
+) -> bool {
+    // Unpack configuration handle, catching null pointers.
+    if config.is_null() {
+        set_last_error("received null configuration handle");
+        return false;
+    }
+    let config = unsafe { &mut (*config).config };
+
+    // Update configuration and return success.
+    config.ignore_unknown_fields_set_to_default();
+    true
+}
+
+/// Whitelists a protobuf message type for use in advanced extensions. If an
```

Comment:
Right, forgot "whitelist" is a trigger word nowadays...

A blocklist doesn't make much sense in this context though, if that's what you meant. According to the Substrait spec, anything is allowed here, so the validator should never throw an error, and it already has to throw a warning by default because it can never guarantee correctness of an extension based on protobuf Any.